iris-mpc-cpu/src/execution/hawk_main/iris_worker.rs

```diff
@@ -0,0 +1,188 @@
+        rsp: oneshot::Sender<Vec<RingElement<u16>>>,
+    },
+}
+
+#[derive(Clone, Debug)]
+pub struct IrisPoolHandle {
+    workers: Arc<Vec<Sender<IrisTask>>>,
+    next_counter: Arc<AtomicU64>,
+}
+
+impl IrisPoolHandle {
+    pub async fn dot_product_pairs(
+        &self,
+        pairs: Vec<(ArcIris, VectorId)>,
+    ) -> Result<Vec<RingElement<u16>>> {
+        let (tx, rx) = oneshot::channel();
+        let task = IrisTask::DotProductPairs { pairs, rsp: tx };
+        self.submit(task, rx).await
+    }
+
+    pub async fn dot_product_batch(
+        &self,
+        query: ArcIris,
+        vector_ids: Vec<VectorId>,
+    ) -> Result<Vec<RingElement<u16>>> {
+        let (tx, rx) = oneshot::channel();
+        let task = IrisTask::DotProductBatch {
+            query,
+            vector_ids,
+            rsp: tx,
+        };
+        self.submit(task, rx).await
+    }
+
+    pub async fn galois_ring_pairwise_distances(
+        &self,
+        input: Vec<Option<(ArcIris, ArcIris)>>,
+    ) -> Result<Vec<RingElement<u16>>> {
+        let (tx, rx) = oneshot::channel();
+        let task = IrisTask::RingPairwiseDistance { input, rsp: tx };
+        self.submit(task, rx).await
+    }
+
+    async fn submit(
+        &self,
+        task: IrisTask,
+        rx: oneshot::Receiver<Vec<RingElement<u16>>>,
+    ) -> Result<Vec<RingElement<u16>>> {
+        let start = Instant::now();
+
+        let _ = self.get_next_worker().send(task);
+        let res = rx.await?;
+
+        histogram!("iris_worker.latency", "histogram" => "histogram")
+            .record(start.elapsed().as_secs_f64());
+        Ok(res)
+    }
+
+    fn get_next_worker(&self) -> &Sender<IrisTask> {
+        // fetch_add() wraps around on overflow
+        let idx = self.next_counter.fetch_add(1, Ordering::Relaxed) as usize;
+        let idx = idx % self.workers.len();
+        &self.workers[idx]
+    }
+}
+
+pub fn init_workers(shard_index: usize, iris_store: SharedIrisesRef<ArcIris>) -> IrisPoolHandle {
+    let core_ids = select_core_ids(shard_index);
+    info!(
+        "Dot product shard {} running on {} cores ({:?})",
+        shard_index,
+        core_ids.len(),
+        core_ids
+    );
+
+    let mut channels = vec![];
+    for core_id in core_ids {
+        let (tx, rx) = crossbeam::channel::unbounded::<IrisTask>();
+        channels.push(tx);
+        let iris_store = iris_store.clone();
+        std::thread::spawn(move || {
+            let _ = core_affinity::set_for_current(core_id);
+            worker_thread(rx, iris_store);
+        });
+    }
+
+    IrisPoolHandle {
+        workers: Arc::new(channels),
+        next_counter: Arc::new(AtomicU64::new(0)),
+    }
+}
+
+fn worker_thread(ch: Receiver<IrisTask>, iris_store: SharedIrisesRef<ArcIris>) {
+    while let Ok(task) = ch.recv() {
+        match task {
+            IrisTask::DotProductPairs { pairs, rsp } => {
+                let pairs = {
+                    let store = iris_store.data.blocking_read();
```

Comment:
Just confirming my understanding: there's no danger of contention here, writes to data only occur when inserting/updating new irises, which happens only for centers towards the end of the pipeline. And not at all during intra-batch and searching.

Reply:
Correct. I’d like some type safety in there though. In fact, I’m trying to move the store exclusively in the pool and remove direct accesses from other places. Continuing this in https://github.com/worldcoin/iris-mpc/pull/1594

Reply:
In the mean time, if there is contention on the lock, that is a bug because that may make the algorithm non-deterministic and unsync the MPC.